File: examples/Clojure/rrbroker.clj
;; Simple request-reply broker

(ns rrbroker
  (:require [zeromq
             [device :as dev]
             [zmq :as zmq]]))

(defn -main []
  (let [context (zmq/zcontext)
        poller (zmq/poller context 2)]
    (with-open [frontend (doto (zmq/socket context :router)
                           (zmq/bind "tcp://*:5559"))
                backend (doto (zmq/socket context :dealer)
                          (zmq/bind "tcp://*:5560"))]
      (zmq/register poller frontend :pollin)
      (zmq/register poller backend :pollin)
      (while (not (.. Thread currentThread isInterrupted))
        (zmq/poll poller)
        (when (zmq/check-poller poller 0 :pollin)
          (loop [part (zmq/receive frontend)]
            (let [more? (zmq/receive-more? frontend)]
              (zmq/send backend part (if more? zmq/send-more 0))
              (when more?
                (recur (zmq/receive frontend))))))
        (when (zmq/check-poller poller 1 :pollin)
          (loop [part (zmq/receive backend)]
            (let [more? (zmq/receive-more? backend)]
              (zmq/send frontend part (if more? zmq/send-more 0))
              (when more?
                (recur (zmq/receive backend))))))))))

File: examples/Rust/rrbroker.rs
#![crate_name = "rrbroker"]

extern crate zmq;

fn main() {
    let context = zmq::Context::new();
    let frontend = context.socket(zmq::ROUTER).unwrap();
    let backend = context.socket(zmq::DEALER).unwrap();
    assert!(frontend.bind("tcp://*:5559").is_ok());
    assert!(backend.bind("tcp://*:5560").is_ok());

    let items = &mut [
        frontend.as_poll_item(zmq::POLLIN),
        backend.as_poll_item(zmq::POLLIN),
    ];

    loop {
        zmq::poll(items, -1).unwrap();
        if items[0].is_readable() {
            loop {
                let message = frontend.recv_msg(0).unwrap();
                let more = if frontend.get_rcvmore().unwrap() {
                    zmq::SNDMORE
                } else {
                    0
                };
                backend.send_msg(message, more).unwrap();
                if more == 0 {
                    break;
                };
            }
        }
        if items[1].is_readable() {
            loop {
                let message = backend.recv_msg(0).unwrap();
                let more = if backend.get_rcvmore().unwrap() {
                    zmq::SNDMORE
                } else {
                    0
                };
                frontend.send_msg(message, more).unwrap();
                if more == 0 {
                    break;
                }
            }
        }
    }
}
